### PR TITLE
Make the CS stricter; fix some bugs / issues

### PR DESCRIPTION
**Pull request by teohhanhui**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | N/A
| License       | MIT
| Doc PR        | N/A

Stricter CS = better code, especially `declare(strict_types=1);`

### CODE REVIEW

File: tests/Bridge/Doctrine/Orm/Filter/DateFilterTest.php [hunk 1]
```diff
@@ -124,14 +126,14 @@ public function provideApplyTestData(): array
             ],
             'after' => [
                 [
-                    'dummyDate' => true,
+                    'dummyDate' => null,
```
Comment (soyuka) [inline tests/Bridge/Doctrine/Orm/Filter/DateFilterTest.php]: why the change here?

File: .php_cs.dist [hunk 2]
```diff
@@ -6,11 +6,43 @@ $finder = PhpCsFixer\Finder::create()
...
+            'throw',
+            'use',
+        ],
+        'no_unreachable_default_argument_value' => true,
+        'no_useless_else' => true,
+        'no_useless_return' => true,
         'ordered_imports' => true,
+        // 'phpdoc_add_missing_param_annotation' => [
+        //     'only_untyped' => false,
+        // ],
```
Comment (soyuka) [inline .php_cs.dist]: should we remove those?

File: tests/Bridge/Doctrine/Orm/Filter/DateFilterTest.php (hunk 1 above)
Comment (teohhanhui) [inline tests/Bridge/Doctrine/Orm/Filter/DateFilterTest.php]: Because this is `$nullManagement`, which can only be `string` or `null`, not `bool`.

File: .php_cs.dist (hunk 2 above)
Comment (teohhanhui) [inline .php_cs.dist]: I plan to get these in next... :smile:

File: .php_cs.dist [hunk 3]
```diff
@@ -6,11 +6,43 @@ $finder = PhpCsFixer\Finder::create()
...
+            'syntax' => 'short',
+        ],
+        'braces' => [
+            'allow_single_line_closure' => true,
+        ],
+        'declare_strict_types' => true,
+        'modernize_types_casting' => true,
+        // 'native_function_invocation' => true,
+        'no_extra_consecutive_blank_lines' => [
+            'break',
```
Comment (teohhanhui) [inline .php_cs.dist]: This one may be a bit controversial, but I think it's good that it encourages us to not have complex logic nested inside a switch-case, or even avoid switch-case altogether.

Comment (dunglas) [inline .php_cs.dist]: I agree with the intent, but it makes the code (even simple statements) less readable (IMO).

File: src/Bridge/Symfony/Routing/RouteNameResolver.php [hunk 4]
```diff
@@ -40,7 +42,7 @@ public function getRouteName(string $resourceClass, bool $collection): string
             $operation = $route->getDefault(sprintf('_api_%s_operation_name', $operationType));
             $methods = $route->getMethods();
 
-            if ($resourceClass === $currentResourceClass && null !== $operation && (empty($methods) || in_array('GET', $methods))) {
+            if ($resourceClass === $currentResourceClass && null !== $operation && (empty($methods) || in_array('GET', $methods, true))) {
```
Comment (dunglas) [inline src/Bridge/Symfony/Routing/RouteNameResolver.php]: The call to `empty` can be replaced by the `!` operator.

**Comment by dunglas** [comment]
Awesome. Can you just fix StyleCI ? Putting the `declare` line before the license header should do the trick.

**Comment by teohhanhui** [comment]
> Can you just fix StyleCI ? Putting the declare line before the license header should do the trick.

No. StyleCI is buggy here. See FriendsOfPHP/PHP-CS-Fixer#2590

Also, declare should go after file-level docblock, according to PSR-12 draft.

File: .php_cs.dist (hunk 3 above)
Comment (teohhanhui) [inline .php_cs.dist]: Yeah, personally I prefer the blank line too, but for the sake of consistency...

File: src/Bridge/Symfony/Routing/RouteNameResolver.php (hunk 4 above)
Comment (teohhanhui) [inline src/Bridge/Symfony/Routing/RouteNameResolver.php]: I really don't like such implicit conversions... And I don't like `empty` too...

My preference would be `0 === count($methods)`, but I understand there is a  performance overhead (how much?) for the function call.

**Comment by dunglas** [comment]
So what do we do in the meantime? Disable StyleCI? Or disable the header check?

File: src/Bridge/Symfony/Routing/RouteNameResolver.php (hunk 4 above)
Comment (dunglas) [inline src/Bridge/Symfony/Routing/RouteNameResolver.php]: I prefer to keep in sync with what we do in Symfony (we remove `empty`, `count` and so on).

File: .php_cs.dist (hunk 3 above)
Comment (dunglas) [inline .php_cs.dist]: Consistency with what?

**Comment by teohhanhui** [comment]
Disabling the header check sounds reasonable. But it might escape our watchful eyes. :smile:

**Comment by dunglas** [comment]
We'll restore the check when PHP CS Fixer will be updated.

File: .php_cs.dist (hunk 3 above)
Comment (teohhanhui) [inline .php_cs.dist]: We cannot enforce a blank line before/after `break;`, but we can enforce `break;` always sticking to the next `case` statement:

```php
switch ($str) {
    case 'A':
        $this->doSomething();

        $this->doSomethingElse();

        break;
    case 'B':
        $this->doSomethingAwesome();
        break;
}
```

Comment (dunglas) [inline .php_cs.dist]: But we can ignore this rule too and check it manually (it's not a big deal after all).

Comment (teohhanhui) [inline .php_cs.dist]: Do you need me to revert this? I'll have to do it manually...

Comment (dunglas) [inline .php_cs.dist]: Don't worry about that, keep it as is.

**Comment by dunglas** [comment]
StyleCI config updated. Can you rebase?

File: .php_cs.dist (hunk 2 above)
Comment (teohhanhui) [inline .php_cs.dist]: @dunglas I want to get this in too... But this is gonna require quite a bit of manual work. :laughing:

File: .php_cs.dist [hunk 5]
```diff
@@ -6,11 +6,43 @@ $finder = PhpCsFixer\Finder::create()
...
+        '@Symfony:risky' => true,
+        'array_syntax' => [
+            'syntax' => 'short',
+        ],
+        'braces' => [
+            'allow_single_line_closure' => true,
+        ],
+        'declare_strict_types' => true,
+        'modernize_types_casting' => true,
+        // 'native_function_invocation' => true,
```
Comment (teohhanhui) [inline .php_cs.dist]: What do you guys think about this? I remember @dunglas talking about this before... We'll need to do Blackfire profiling for this to see the performance difference.

**Comment by teohhanhui** [comment]
Rebased. But StyleCI still has a problem with `return null;` when Symfony has already changed its CS in view of the introduction of the `void` return type. I suspect they are using outdated PHP-CS-Fixer? :disappointed:

**Comment by teohhanhui** [comment]
It could make sense for us to disable StyleCI (at least for now?), because frankly I don't see what's the added value of having it.

**Comment by dunglas** [comment]
Are you sure Symfony has changed this CS?

**Comment by dunglas** [comment]
Anyway StyleCI is useless since we run our own PHP CS Fixer instance. I've just disabled it. Can you rebase again?

**Comment by soyuka** [comment]
Styleci would make sense if it prevented the other CI from running on failure.

**Comment by teohhanhui** [comment]
Yes, since symfony/symfony-docs#6614

> * Use `return null;` when a function explicitly returns `null` values and use `return;` when the function returns `void` values;
>
> http://symfony.com/doc/current/contributing/code/standards.html

**Comment by teohhanhui** [comment]
The only merit I see for StyleCI is the nice output, and the ability to send a PR. :laughing:

**Comment by dunglas** [comment]
Thank you @teohhanhui
